src/main/java/frc/robot/commands/SpeakerCommand.java

```diff
@@ -1,26 +1,80 @@
 package frc.robot.commands;
 
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.kinematics.ChassisSpeeds;
+import edu.wpi.first.wpilibj.DriverStation;
 import edu.wpi.first.wpilibj2.command.Command;
+import frc.robot.constants.ShooterConstants;
 import frc.robot.subsystems.IndexerSubsystem;
 
 public class SpeakerCommand extends Command {
 
     private final IndexerSubsystem indexerSubsystem;
+    private final ChassisSpeeds fieldRelativeSpeeds;
+    private final Pose2d swervePose;
+
 
     /**
      * Command to shoot for the speaker once there is a note in the intake
      * @param indexerSubsystem Instance of {@link IndexerSubsystem}
      */
 
-    public SpeakerCommand(IndexerSubsystem indexerSubsystem) {
+    public SpeakerCommand(IndexerSubsystem indexerSubsystem, ChassisSpeeds fieldRelativeSpeeds, Pose2d swervePose) {
         this.indexerSubsystem = indexerSubsystem;
-
+        this.fieldRelativeSpeeds = fieldRelativeSpeeds;
+        this.swervePose = swervePose;
         addRequirements(indexerSubsystem);
     }
 
     @Override
     public void execute() {
-        indexerSubsystem.rotateAllWheelsPercent(1);
+        double robotXSpeed = fieldRelativeSpeeds.vxMetersPerSecond;
+        double noteSpeedX = robotXSpeed + ShooterConstants.shooterNoteSpeedX;
+
+        double shootingDistanceFromSpeaker = noteSpeedX * ShooterConstants.timeToSpeakerHeightWithGravity;
+
+        double robotDistanceFromSpeaker = Math.sqrt(
+                Math.pow(
+                        (
+                            (DriverStation.getAlliance().get() == DriverStation.Alliance.Red)?
+                                    ShooterConstants.speakerPoseRed.getX():
+                                    ShooterConstants.speakerPoseBlue.getX()
+                        ) - swervePose.getX(),
+                        2
+                )
+                + Math.pow(
+                        (
+                                (DriverStation.getAlliance().get() == DriverStation.Alliance.Red)?
+                                        ShooterConstants.speakerPoseRed.getY():
+                                        ShooterConstants.speakerPoseBlue.getY()
+                        ) - swervePose.getY(),
+                        2
+                )
+        );
+
+        double difference = robotDistanceFromSpeaker - shootingDistanceFromSpeaker;
+
+        if (difference > 0) {
+            shootingDistanceFromSpeaker += noteSpeedX * ShooterConstants.shootDelayTime;
+        } else {
+            shootingDistanceFromSpeaker -= noteSpeedX * ShooterConstants.shootDelayTime;
```

Comment:
Im assuming this delay time if accounting for the delay to move the note from the indexer to the shooter. If it is not, that time needs to be considered for.